keps/sig-node/0036-20190118-ForceEviction.md

```diff
@@ -0,0 +1,178 @@
+In addtion, the containers status for the Pod are still running since kubelet cannot update it,
+which blocks attach_detach controller to detach volumes. The current taint-based eviction actually
+cannot give any benefit in the sense that Pod can only be succesfully evicted after Kubelet comes
+back healthy, in which case there is no need of Pod eviction at all.
+
+
+### Goals
+
+This proposal is trying to investigate the failure handling issues for pods
+and their volumes with a goal of improving system reliability and failure recovery capabilities.
+Among different kinds of failures, node shut down is the main user case this proposal tries
+to address. In this case, we are sure that all containers are stopped, and all mounts are gone
+so that it is safe to take some recovery actions more aggresively without waiting for kubelet correspondence.  
+
+to delete the Pods from API server without waiting for kubelet's acknoledgement.
+Volumes can also be safely detached since mounts no longer exist.
+
+### Non-Goals
+
+This proposal is not trying to address the recovery issue for all types of failures. For example for network
+partition issue, because this is no source of truth about whether Pods are running or not, it is not possible
+for controller decide what recovery actions should take based on the node status.
+
+## Proposal
+
+Currently, Kubernetes already supports four types of taint effects:
+- NoSchedule: Pods that do not tolerate this taint are not scheduled on the node.
+- PreferNoSchedule: Kubernetes avoids scheduling Pods that do not tolerate this taint onto the node.
+- NoExecute: Pod is evicted from the node if it is already running on the node, and is not scheduled
+onto the node if it is not yet running on the node.
+
+We propose to add a new effect, TaintEffectForceEviction. For a Pod that cannot tolerant this taint,
+it will be force deleted from the node with zero grace period which means that Pod can be deleted immediately
+without waiting for Kubelet to responde. One important use case of this taint is node shut down case. If node
+controller or some external controller detact that node is powered off, the controller could add the following
+taint to the node.
+
+
+```
+taint := &v1.Taint{
+	Key:    nodeShutdown,
+	Value:  "true",
+	Effect: v1.TaintEffectForceEviction,
+}
+```
+
+After the node is tainted, node lifecycle controller will trigger to forcefully delete all the Pods from the node.
+If Kubelet comes back healthy in the middle of force eviction process, there might be race condition between kubelet
+starting Pods while they are being deleted since force deletion does not require kubelet coordination any more.
+To avoid race condition between controller and kubelet, kubelet also requires some additional handling when node
+is tainted which is explained in more details below.
+
+For node lifecycle controller,  we plan to add the following eviction logic:
```

Comment:
The original design for pod safety called for a user specified duration distinct from the grace period so that we are defining a toleration policy that can be tuned to workloads. Where is that in this proposal?

Reply:
I’m not sure to understand here, do you suggest adding a grace period before starting to evict based on this new effect ? If that’s what you meant then yes we can add this, One of the usecases where they’d want to do this is quick reboots i imagine, where users would set a proper grace period + tolerate the taint.

Reply:
The proposal here is to force delete the pod with 0 grace period. The use case of this taint is different from the normal taint-based eviction. If the node is tainted with this ForceEviction effect by a controller or admin, it means that the pods are no longer running, system want to have a fast recovery from this situation. The most common scenario would be node is powered off. Whoever taint the node should understand the consequence of this taint which will evict the pods forcefully with 0 grace period. The design also covers the case of node is coming back during eviction. Basically we will avoid starting new pod on the node if there is ForceEviction taint.